lib/SILOptimizer/Transforms/CopyPropagation.cpp

```diff
@@ -0,0 +1,908 @@
+  case SILInstructionKind::ClassMethodInst:
+  case SILInstructionKind::CopyBlockInst:
+  case SILInstructionKind::CopyValueInst:
+  case SILInstructionKind::DebugValueInst:
+  case SILInstructionKind::ExistentialMetatypeInst:
+  case SILInstructionKind::FixLifetimeInst:
+  case SILInstructionKind::SelectEnumInst:
+  case SILInstructionKind::SetDeallocatingInst:
+  case SILInstructionKind::StoreWeakInst:
+  case SILInstructionKind::ValueMetatypeInst:
+    return false;
+
+  // Escape the value. The lifetime must already be enforced via something like
+  // fix_lifetime.
+  case SILInstructionKind::RefToRawPointerInst:
+  case SILInstructionKind::RefToUnmanagedInst:
+  case SILInstructionKind::RefToUnownedInst:
+  case SILInstructionKind::UncheckedBitwiseCastInst:
+  case SILInstructionKind::UncheckedTrivialBitCastInst:
+    return false;
+
+  // Dynamic dispatch without capturing self.
+  case SILInstructionKind::ObjCMethodInst:
+  case SILInstructionKind::ObjCSuperMethodInst:
+  case SILInstructionKind::SuperMethodInst:
+  case SILInstructionKind::WitnessMethodInst:
+    return false;
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// CopyPropagationState: shared state for the pass's analysis and transforms.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// LiveWithin blocks have at least one use and/or def within the block, but are
+/// not LiveOut.
+///
+/// LiveOut blocks are live on at least one successor path.
+///
+/// Dead blocks have not yet been discovered by the liveness computation.
+enum IsLive_t { LiveWithin, LiveOut, Dead };
+
+/// Liveness information, which is produced by step #2, computeLiveness, and
+/// consumed by step #3, findOrInsertDestroys.
+class LivenessInfo {
+  // Map of all blocks in which current def is live. True if it is also liveout.
+  DenseMap<SILBasicBlock *, bool> liveBlocks;
+  // Set of all "interesting" users in this def's live range and whether their
+  // used value is consumed. (Non-consuming uses within a block that is already
+  // known to be live are uninteresting.)
+  DenseMap<SILInstruction *, bool> users;
+  // Original points in the CFG where the current value was consumed or
+  // destroyed.
+  typedef SmallSetVector<SILBasicBlock *, 8> BlockSetVec;
+  BlockSetVec originalDestroyBlocks;
+
+public:
+  bool empty() {
+    assert(!liveBlocks.empty() || users.empty());
+    return liveBlocks.empty();
+  }
+
+  void clear() {
+    liveBlocks.clear();
+    users.clear();
+    originalDestroyBlocks.clear();
+  }
+
+  IsLive_t isBlockLive(SILBasicBlock *bb) const {
+    auto liveBlockIter = liveBlocks.find(bb);
+    if (liveBlockIter == liveBlocks.end())
+      return Dead;
+    return liveBlockIter->second ? LiveOut : LiveWithin;
+  }
+
+  void markBlockLive(SILBasicBlock *bb, IsLive_t isLive) {
+    assert(isLive != Dead && "erasing live blocks isn't implemented.");
+    liveBlocks[bb] = isLive;
```

Comment:
implicitly converting an enum to a bool.
Better: liveBlocks[bb] = (isLive == LiveOut);

Reply:
Of course. Done.